File: dropwizard/src/main/java/com/yammer/dropwizard/jersey/OauthTokenProvider.java
package com.yammer.dropwizard.jersey;

import com.google.common.base.Optional;
import com.sun.jersey.api.model.Parameter;
import com.sun.jersey.core.spi.component.ComponentContext;
import com.sun.jersey.core.spi.component.ComponentScope;
import com.sun.jersey.spi.inject.Injectable;
import com.sun.jersey.spi.inject.InjectableProvider;
import com.yammer.dropwizard.BearerToken;

import javax.ws.rs.ext.Provider;

// TODO: 11/14/11 <coda> -- test OauthTokenProvider
// TODO: 11/14/11 <coda> -- document OauthTokenProvider

@Provider
public class OauthTokenProvider implements InjectableProvider<BearerToken, Parameter> {
    @Override
    public ComponentScope getScope() {
        return ComponentScope.PerRequest;
    }

    @Override
    public Injectable<?> getInjectable(ComponentContext ic,
                                       BearerToken a,
                                       Parameter c) {
        if (c.getParameterClass().isAssignableFrom(Optional.class)) {
            return new OauthTokenInjectable(a.value() + ' ');
        }
        return null;
    }
}

File: dropwizard-scala_2.9.1/src/main/scala/com/yammer/dropwizard/providers/OauthTokenProvider.scala
package com.yammer.dropwizard.providers

import javax.ws.rs.core.HttpHeaders
import javax.ws.rs.ext.Provider
import com.sun.jersey.api.core.HttpContext
import com.sun.jersey.api.model.Parameter
import com.sun.jersey.spi.inject.InjectableProvider
import com.sun.jersey.core.spi.component.{ComponentScope, ComponentContext}
import com.sun.jersey.server.impl.inject.AbstractHttpContextInjectable
import com.yammer.dropwizard.BearerToken

@Provider
class OauthTokenProvider extends InjectableProvider[BearerToken, Parameter] {
  def getInjectable(ic: ComponentContext, a: BearerToken, c: Parameter) =
    if (c.getParameterClass.isAssignableFrom(classOf[Option[String]])) {
      new OauthTokenInjectable(a.prefix() + " ")
    } else {
      null
    }

  def getScope = ComponentScope.PerRequest
}

class OauthTokenInjectable(prefix: String) extends AbstractHttpContextInjectable[Option[String]] {
  def getValue(c: HttpContext) = {
    val header = Option(c.getRequest.getHeaderValue(HttpHeaders.AUTHORIZATION))
    header.collect {
      case token if token.startsWith(prefix) => token.substring(prefix.length())
    }
  }
}
